### PR TITLE
Configure Swagger Runner with the application & /node endpoints - Closes #918, #924

### LINKED ISSUE #899: Swagger Integration with Core
Belongs to: #225 

Right now we are managing schema of all our APIs in json-schema formats within app. And then defining same schema again in swagger. That's make redundant code. 

Need changes in following: 
1. API loading and endpoint registration
2. Parameter validations 
3. Response validations 

So the definition in Swagger should only be used. 

Related Task List:

- [x] Configure Swagger Runner with the application #918
- [x] Add Swagger Based Response Validation in Test Cases #919
- [x] Move existing express middleware to swagger-runner pipelines #1011 
- [x] Standarisation of swagger responses #674 

Endpoint Task List:

- [x] accounts #920 
- [x] blocks #921 
- [x] dapps #922 
- [x] delegates #923 
- [x] multisignatures
  - [x] multisignature-accounts #1116 
  ~~- [ ] multisignature-pending~~ https://github.com/LiskHQ/lisk/issues/899#issuecomment-350706885
- [x] node #924 
- [x] peers #925 
- [x] signatures #926 
- [x] transactions #927 
- [x] votes #928 
- [x] voters #929

### LINKED ISSUE #918: Configure Swagger Runner with the application
Parent: #899

### LINKED ISSUE #924: Refactor /node endpoints to use swagger runner interface
Parent: #899

### PR DESCRIPTION
**Pull request by nazarhussain**
Closes #918 
Closes #924 

Parent: #899

### CODE REVIEW

File: api/controllers/node.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	modules.node.shared.getStatus(null, function (err, data){
+		res.send(data);
+	});
+}
+
+module.exports = {
+	getConstants: getConstants,
+	getStatus: getStatus,
+	bind: Controller
+};
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply api/controllers/node.js]: Will be taken care of in next commit.

File: helpers/swagger_security_handlers.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+
+
+module.exports = {
+    
+};
```
Comment (Isabello) [inline helpers/swagger_security_handlers.js]: Is this place holder required?

File: test/common/apiHelpers.js [hunk 3]
```diff
@@ -96,11 +96,11 @@ function getForgingStatus (params, cb) {
 }
 
 function getNodeConstants (cb) {
-	http.get('/api/node/constants', httpCallbackHelper.bind(null, cb));
+	http.get('/LiskHQ/Lisk/1.0.18/node/constants', httpCallbackHelper.bind(null, cb));
```
Comment (Isabello) [inline test/common/apiHelpers.js]: Can we change this base path to a param from upstream?

File: helpers/swagger_security_handlers.js (hunk 2 above)
Comment (nazarhussain) [reply helpers/swagger_security_handlers.js]: Its a placeholder to move code of `httpApi.middleware.applyAPIAccessRules` to swagger pipe stack.

File: test/common/apiHelpers.js (hunk 3 above)
Comment (nazarhussain) [reply test/common/apiHelpers.js]: Yes we can do it. I attached validated swagger to the app for this purpose already. Will see how to get rid of the base path from tests.

File: api/controllers/peers.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+		version: req.swagger.params.version.value,
+		broadhash: req.swagger.params.broadhash.value,
+		height: req.swagger.params.height.value,
+		limit: req.swagger.params.limit.value,
+		offset: req.swagger.params.offset.value,
+		orderBy: req.swagger.params.sort.value
+	};
+
+	// Remove filters with null values
+	filters = _.pickBy(filters, _.identity);
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply api/controllers/peers.js]: Swagger params objects `req.swagger.params` came up with all parameters defined in definition. Even if not passed by the user to the request. It treated these as null or default value defined in specification. We need to remove all such params with null values, means not specified by end user.

File: config/swagger.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		swagger: swagger,
+		enforceUniqueOperationId: true,
+		startWithErrors: false,
+		startWithWarnings: true
+	  };
+
+	  SwaggerRunner.create(swaggerConfig, function(err, runner) {
+		
+		if (err) {
+			// Some error occured in configuring the swagger
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply config/swagger.js]: Yes its a string here. Grammar fixed.

File: schema/swagger.yml [hunk 6]
```diff
@@ -0,0 +1,1843 @@
...
+      - in: query
+        name: secondPublicKey
+        description: publicKey to query
+        type: string
+      - in: query
+        name: sort
+        description: fields to sort on to query
+        required: false
+        type: string
+      responses:
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply schema/swagger.yml]: Will be added when will work on this endpoint.

File: package.json [hunk 7]
```diff
@@ -65,6 +65,7 @@
     "socketcluster-client": "=5.3.0",
     "sodium": "LiskHQ/node-sodium#716de00",
     "strftime": "=0.10.0",
+    "swagger-node-runner": "^0.7.3",
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply package.json]: Done.

File: schema/swagger.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Isabello) [reply schema/swagger.yml]: I believe we discard 204 because it erases the error message

File: config/swagger.js [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+		});
+				
+		// install middleware
+		swaggerExpress.register(app);
+
+		// To be used in test cases or getting configuration runtime
+		app.swaggerRunner = runner;
+
+		// Successfully mounted the swagger runner
+		cb(null, { swaggerRunner: runner });
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply config/swagger.js]: What dod you refer in this comment? I can't understand.

File: config/swagger.js [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * Configure swagger node runner with the app.
+ * It loads swagger specification and map every thing with an active express app
+ * @requires swagger-node-runner
+ * @requires path
+ * @requires fs
+ * @requires js-yaml
+ * @module config:swagger
+ * @param {object} app - An express app to which map the swagger details
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply config/swagger.js]: Done.

File: api/fittings/custom_param_formats.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+	// Register lisk formats with swagger
+	Object.keys(z_schema.formatsCache).forEach(function (formatName) {
+		// Extend swagger validator with our formats
+		validator.constructor.registerFormat(formatName, z_schema.formatsCache[formatName]);
+	});
+
+	return function custom_params_formats (context, cb) {
+		cb();
+	};
+};
```
[the remark this answers is not in the document]
Comment (nazarhussain) [reply api/fittings/custom_param_formats.js]: Done.

File: schema/swagger.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (MaciejBaj) [reply schema/swagger.yml]: We just merged voters endpoint which uses it. Also, functional tests helpers are adapted to handle an empty response in this case.

File: schema/swagger.yml [hunk 11]
```diff
@@ -0,0 +1,1843 @@
...
+        example: 7199d4b67c3575d5f99d1c29436a02977eeb01a7
+      version:
+        type: string
+        example: v0.8.0
+      nethash:
+        type: string
+        example: ed14889723f24ecc54871d058d98ce91ff2f973192075c0155ba2b7b70ad2511
+      supply:
+        type: string
+        example: 10575384500000000
```
Comment (MaciejBaj) [inline schema/swagger.yml]: Looks like a number to me

File: schema/swagger.yml [hunk 12]
```diff
@@ -0,0 +1,1843 @@
...
+        type: integer
+        example: 2
+      lifetime:
+        type: integer
+        example: 72
+
+
+# Added by API Auto Mocking Plugin
+host: virtserver.swaggerhub.com
+basePath: /LiskHQ/Lisk/1.0.18
```
Comment (MaciejBaj) [inline schema/swagger.yml]: Empty line missing

File: test/functional/shared.js [hunk 13]
```diff
@@ -196,10 +197,26 @@ function MultisigScenario (size, amount) {
 	this.amount = amount || 100000000000;
 }
 
+function validateParameterResponse (res, paramNames, message) {
+	node.expect(res.body).to.have.property('message').to.equal(message);
+}
+
+function validateSwaggerParameterResponse (res, paramNames, message) {
```
Comment (MaciejBaj) [inline test/functional/shared.js]: I suggest to change a name for sth like `expectSwaggerErrorResponse`

File: config/swagger.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Isabello) [reply config/swagger.js]: spacing standards

File: schema/swagger.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (Isabello) [reply schema/swagger.yml]: fine with me

File: test/functional/shared.js (hunk 13 above)
Comment (Isabello) [reply test/functional/shared.js]: agree with this naming

File: test/node.js [hunk 14]
```diff
@@ -438,7 +447,7 @@ node.initApplication = function (cb, initScope) {
 			},
 			network: function (cb) {
 				// Init with empty function
-				cb(null, {io: {sockets: {emit: function () {}}}});
+				cb(null, {io: {sockets: {emit: function () {}}}, app: require('express')()});
```
Comment (MaciejBaj) [inline test/node.js]: Are you sure that we need it and it will not conflict with currently running tests?

Comment (nazarhussain) [reply test/node.js]: We need an express app to mount the swagger middleware. Without it we can't test the swagger files properly. Yes all tests are passing.
